### DISCUSSION TITLE
throwing instead of assert for `const&`

### DISCUSSION THREAD

**Discussion by tawmoto** [nlohmann/json#2466] [Q&A] [answered]
Hello,

It seems that in certain cases an ASSERT is executed instead of throwing an exception, this causes unit tests to fail on the debug and the program crashes. I could reproduce it with a small example but now it works :(, I can reproduce it anytime with my project.

```
Json j = "{ \"happy\": true, \"pi\": 3.141 }"_json;

try
{
     auto val = j["garbage"].get<int>();
}
catch (const Json::exception& ex)
{
    auto str = ex.what(); // the exception is nicely caught here.
}
```

Now same code but using a lambda
```
Json j = "{ \"happy\": true, \"pi\": 3.141 }"_json;

auto getVal = [](const Json& j) {
        auto val = j["garbage"].get<int>();
        return val;
    };

try
{
    auto val = j["garbage"].get<int>();
}
catch (const Json::exception& ex)
{
    auto str = ex.what();
}
```
Nothing happens, the program just crashes, because it goes to this line: (but my json is a reference, not a pointer)
https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp#L20218

Sometimes it goes to:
https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp#L20204
I don't understand why, because I am making a copy, 

My question is: why is the `const_reference` called instead of why not throw instead of ASSERT?

In my original code I have smth like
```
struct FilterListUpdateData
{
    int version{0};
   // other data...
};

FilterListUpdateData filterListInfo;
filterListInfo.version = json["garbage"].get<int>(); // crashes everytime because of the ASSERT
```

Any advices? Maybe should I use `at()` instead of `[]` everywhere?
Thank you

**Comment by nlohmann** [comment, marked as answer]
The first snippet calls `operator[]` on a non-const `json` value, whereas the latter does so for a const `json` value. The behavior is different, just as you observed. Please use `at` if you want checked access which always throws in case of a missing key.

Documentation:
- https://nlohmann.github.io/json/features/element_access/unchecked_access/
- https://nlohmann.github.io/json/api/basic_json/operator%5B%5D/
- https://nlohmann.github.io/json/api/basic_json/at/
